test/upgrade/testing_operations_test.go

```diff
@@ -0,0 +1,308 @@
+				return empty
+			}
+			return fmt.Sprintf(mf.skipped, args...)
+		},
+		starting: func(args ...interface{}) string {
+			return fmt.Sprintf(mf.starting, args...)
+		},
+		element: func(args ...interface{}) string {
+			return fmt.Sprintf(mf.element, args...)
+		},
+	}
+}
+
+func (tt *texts) append(msgs ...string) *texts {
+	for _, msg := range msgs {
+		if msg == "" {
+			continue
+		}
+		tt.elms = append(tt.elms, msg)
+	}
+	return tt
+}
+
+func completeSuiteExample(fp failurePoint) upgrade.Suite {
+	suite := upgrade.Suite{
+		Tests: upgrade.Tests{
+			PreUpgrade: []upgrade.Operation{
+				serving.tests.preUpgrade, eventing.tests.preUpgrade,
+			},
+			PostUpgrade: []upgrade.Operation{
+				serving.tests.postUpgrade, eventing.tests.postUpgrade,
+			},
+			PostDowngrade: []upgrade.Operation{
+				serving.tests.postDowngrade, eventing.tests.postDowngrade,
+			},
+			ContinualTests: []upgrade.BackgroundOperation{
+				serving.tests.continual, eventing.tests.continual,
+			},
+		},
+		Installations: upgrade.Installations{
+			Base: []upgrade.Operation{
+				serving.installs.stable, eventing.installs.stable,
+			},
+			UpgradeWith: []upgrade.Operation{
+				serving.installs.head, eventing.installs.head,
+			},
+			DowngradeWith: []upgrade.Operation{
+				serving.installs.stable, eventing.installs.stable,
+			},
+		},
+	}
+	return enrichSuiteWithFailures(suite, fp)
+}
+
+func emptySuiteExample() upgrade.Suite {
+	return upgrade.Suite{
+		Tests:         upgrade.Tests{},
+		Installations: upgrade.Installations{},
+	}
+}
+
+func enrichSuiteWithFailures(suite upgrade.Suite, fp failurePoint) upgrade.Suite {
+	steps := createSteps(suite)
+	for i, st := range steps {
+		for j, op := range st.ops.ops {
+			if fp.step == i+1 && fp.element == j+1 {
+				op.fail(fp.step == 3)
+			}
+		}
+	}
+	return recreateSuite(steps)
+}
+
+func recreateSuite(steps []*step) upgrade.Suite {
+	suite := &upgrade.Suite{
+		Tests:         upgrade.Tests{},
+		Installations: upgrade.Installations{},
+	}
+	for _, st := range steps {
+		st.updateSuite(st.ops, suite)
+	}
+	return *suite
+}
+
+func (o operation) Name() string {
+	if o.op != nil {
+		return o.op.Name()
+	} else {
+		return o.bg.Name()
+	}
+}
+
+func (o *operation) fail(setupFail bool) {
+	failureTestingMessage := "This error is expected to be seen. Upgrade suite should fail."
+	testName := fmt.Sprintf("FailingOf%s", o.Name())
+	if o.op != nil {
+		prev := o.op
+		o.op = upgrade.NewOperation(testName, func(c upgrade.Context) {
+			handler := prev.Handler()
+			handler(c)
+			c.T.Error(failureTestingMessage)
+			c.Log.Error(failureTestingMessage)
+		})
+	} else {
+		prev := o.bg
+		o.bg = upgrade.NewBackgroundOperation(testName, func(c upgrade.Context) {
+			handler := prev.Setup()
+			handler(c)
+			if setupFail {
+				c.T.Error(failureTestingMessage)
+				c.Log.Error(failureTestingMessage)
+			}
+		}, func(bc upgrade.BackgroundContext) {
+			waitForStopSignal(bc, prev.Name(), func(sig upgrade.StopEvent) interface{} {
```

Comment:
I see adding this method `waitForStopSignal` as an improvement. But it's still pretty hard to figure how the user should write his "handler".
They need to implement the function which accepts upgrade.BackgroundContext, then  they have to reuse `waitForStopSignal` in a specific way and pass a third function/handler that accepts `upgrade.StopEvent` which returns a general `interface{}`.
AFAIK, usually the content of `func( bc upgrade.BackgroundContext){}` will run continuously and the user has to have some kind of a loop there. This waitForStopSignal now holds the loop and it's not clear where the user would put his own loop.
IMO, it would help to have a test that is close to real world: have a short loop doing some actions and then waiting for the signal, finishing the background test and sending a signal back. Perhaps it would make more clear what can be simplified - what is really required here and what can be removed.

Reply:
It's also not clear how the "setup" and "handler" would communicate with each other. Let's say the setup phase needs to setup some clients/service that will be used later in the background test. How would I do that? An example closer to real world could clarify that.

Reply:
I presume that users will implement using `NewBackgroundVerification` method. See example at: https://github.com/cardil/pkg/blob/c582e40/test/upgrade/suite_examples_test.go#L150-L159 or with combination of `NewBackgroundOperation` and `WaitForStopEvent` (example at: https://github.com/cardil/pkg/blob/c582e40/test/upgrade/suite_examples_test.go#L105-L123).

I hadn't added any way to pass variables between `Setup` and `Handler` methods, as that would require to be a map-like thing. I envision that most communication would be done by users in their code. Also, I hadn't written such example in tests to not obscure the image of upgrade suite.  Below is an example of such communication done with simple variables:

```golang
// file: knative.dev/serving/test/e2e/upgrade/continual.go
package upgrade

import (
  pkgupgrade "knative.dev/pkg/test/upgrade"
  "knative.dev/serving/test/e2e"
)

func ContinualTest() pkgupgrade.BackgroundOperation {
  var (
    namespace string
    service   string
  )
  return upgrade.NewBackgroundVerification("Serving continual test",
    func(c upgrade.Context) {
      namespace = e2e.GenerateTestName("serving-upgrades")
      service = e2e.GenerateTestName("prober")
      e2e.Clients.Knative.Serving.Namespace(namespace).Create(Service{
         // [..]
      })
    },
    func(c upgrade.Context) {
      ksvc := e2e.Clients.Knative.Serving.Namespace(namespace).Get(service)
      // do a checking
    },
  )
}
```